### PR TITLE
Vectorize sigmoid

### PR DESCRIPTION
**Pull request by cpuhrsch**
This PR ports the vectorization of sigmoid to also enable better performance for non-contiguous arrays. Detailed timings will follow shortly.

### CODE REVIEW

**Comment by cpuhrsch** [comment]
Timing were retrieved [using this benchmark](https://github.com/pytorch/benchmark/blob/master/timing/python/benchmarks/cpu_unary_benchmark.py).

The command run was

```
OMP_NUM_THREADS=1 numactl --membind=0 --cpubind=0 taskset -c 0 python run.py --include CPUUnaryBench --benchmark-min-time 2 --benchmark-warmup-repetitions 2 --benchmark-repetitions 3
```

This is a single core benchmark.

```
framework                                Torch   Torch Master     Ratio  Better
cont  trans function     dim mag
False False ('sigmoid',) 3   1        6.930478       6.567699  0.947655   False
                             3       13.235672      29.237036  2.208957    True
                             6     8351.410715   30095.457062  3.603638    True
                             7    85592.009994  308379.842767  3.602905    True
      True  ('sigmoid',) 3   1        7.514903       6.595705  0.877683   False
                             3       18.093501      29.376038  1.623568    True
                             6     8895.611246   30055.028340  3.378636    True
                             7    91487.099854  307583.449497  3.362042    True
True  False ('sigmoid',) 3   1        6.942165       6.119649  0.881519   False
                             3       12.475567       7.145404  0.572752   False
                             6     1647.585599    1244.441353  0.755312   False
                             7    16909.896146   13191.346432  0.780096   False
      True  ('sigmoid',) 3   1        7.536675       6.587368  0.874042   False
                             3       17.715659      28.759583  1.623399    True
                             6     8135.648723   29658.940342  3.645553    True
                             7    79345.416158  303789.017412  3.828690    True
```

There are significant gains for the non-contiguous cases if the Tensor is larger than 10 elements. However, there is a regression for the regular contiguous case. This needs to be resolved before this can be merged.

**Comment by cpuhrsch** [comment]
We found and mitigated the perf issue and will treat it separately.

These are the new speedups

```
                                                              time_mean
dtype         sizes                       strides
torch.float32 torch.Size([215, 215, 215]) (215, 46225, 1)     16.867260
                                          (3870, 832050, 18)   2.989488
                                          (46225, 215, 1)      1.039506
                                          (832050, 3870, 18)   2.960994
              torch.Size([99, 99, 99])    (176418, 1782, 18)   3.477831
                                          (1782, 176418, 18)   3.087606
                                          (9801, 99, 1)        1.037980
                                          (99, 9801, 1)       14.272740
torch.float64 torch.Size([215, 215, 215]) (215, 46225, 1)      4.597996
                                          (3870, 832050, 18)   2.413990
                                          (46225, 215, 1)      4.718199
                                          (832050, 3870, 18)   2.474521
              torch.Size([99, 99, 99])    (176418, 1782, 18)   2.544542
                                          (1782, 176418, 18)   2.373826
                                          (9801, 99, 1)        4.980543
                                          (99, 9801, 1)        4.441531
```

```
$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                80
On-line CPU(s) list:   0-79
Thread(s) per core:    2
Core(s) per socket:    20
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 79
Model name:            Intel(R) Xeon(R) CPU E5-2698 v4 @ 2.20GHz
Stepping:              1
CPU MHz:               1641.835
CPU max MHz:           3600.0000
CPU min MHz:           1200.0000
BogoMIPS:              4400.92
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              51200K
NUMA node0 CPU(s):     0-19,40-59
NUMA node1 CPU(s):     20-39,60-79
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdseed adx smap xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts
```

Command
```
> 55685  sudo cset shield --exec -- taskset -c 20 python run.py --include CPUUnaryBench --benchmark-min-time 2 --benchmark-warmup-repetitions 2 --benchmark-repetitions 3 --benchmark-filter .*sigmoid.* --benchmark-out /tmp/2
```

Benchmark commit: [d7b07460f401363888e6e5343eee7079b70374c8](https://github.com/pytorch/benchmark/commit/d7b07460f401363888e6e5343eee7079b70374c8)

**Comment by cpuhrsch** [comment]
ROCM build [succeeded separately](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang3.8-rocmnightly-ubuntu16.04-build/1818//console).

File: aten/src/ATen/cpu/vml.h [hunk 1]
```diff
@@ -51,70 +85,82 @@ inline void vrsqrt(scalar_t* out, scalar_t* in, int64_t size) {
...
+IMPLEMENT_VML_BUG(trunc)
+
+#if AT_MKL_ENABLED() && !defined(__APPLE__)
+
+#define IMPLEMENT_VML_MKL(op, mklop)                                         \
   template <>                                                                 \
-  inline void v##op(float* out, float* in, int64_t size) {                    \
-    vms##mklop(size, in, out, VML_HA | VML_FTZDAZ_OFF | VML_ERRMODE_IGNORE); \
+  inline void v##op(float* out, const float* in, int64_t size) {              \
+    vms##mklop(size, in, out, VML_HA | VML_FTZDAZ_OFF | VML_ERRMODE_DEFAULT); \
```
Comment (colesbury) [inline aten/src/ATen/cpu/vml.h]: Why the change? Didn't this cause a lot of warnings to be printed to stderr?

**Review by colesbury — APPROVED** [review]
